modules/core/src/main/java/org/apache/ignite/internal/processors/metastorage/persistence/DmsDataWriterWorker.java

```diff
@@ -81,134 +90,125 @@ public DmsDataWriterWorker(
+                metastorage.writeRaw(localKey(item.key), item.valBytes);
 
-                try {
-                    update = updateQueue.take();
-                }
-                catch (InterruptedException ignore) {
-                }
+            for (int i = 0, len = fullNodeData.hist.length; i < len; i++) {
+                DistributedMetaStorageHistoryItem histItem = fullNodeData.hist[i];
 
-                lock.lock();
+                long histItemVer = fullNodeData.ver.id() + i - (len - 1);
 
-                try {
-                    // process update
-                    if (update instanceof DistributedMetaStorageHistoryItem)
-                        applyUpdate((DistributedMetaStorageHistoryItem)update);
-                    else if (update instanceof DistributedMetaStorageClusterNodeData) {
-                        DistributedMetaStorageClusterNodeData fullNodeData = (DistributedMetaStorageClusterNodeData)update;
+                metastorage.write(historyItemKey(histItemVer), histItem);
+            }
 
-                        metastorage.writeRaw(cleanupGuardKey(), DUMMY_VALUE);
+            metastorage.write(versionKey(), fullNodeData.ver);
 
-                        doCleanup();
+            workerDmsVer = fullNodeData.ver;
 
-                        for (DistributedMetaStorageKeyValuePair item : fullNodeData.fullData)
-                            metastorage.writeRaw(localKey(item.key), item.valBytes);
+            metastorage.remove(cleanupGuardKey());
+        }));
+    }
 
-                        for (int i = 0, len = fullNodeData.hist.length; i < len; i++) {
-                            DistributedMetaStorageHistoryItem histItem = fullNodeData.hist[i];
+    /** */
+    public void removeHistItem(long ver) {
+        updateQueue.offer(newDmsTask(() -> metastorage.remove(historyItemKey(ver))));
+    }
 
-                            long histItemVer = fullNodeData.ver.id() + i - (len - 1);
+    /** */
+    public void cancel(boolean halt) {
+        if (halt)
+            updateQueue.clear();
 
-                            metastorage.write(historyItemKey(histItemVer), histItem);
-                        }
+        updateQueue.offer(new FutureTask<>(() -> STOP));
 
-                        metastorage.write(versionKey(), fullNodeData.ver);
+        U.join(runner(), log);
+    }
 
-                        workerDmsVer = fullNodeData.ver;
+    /** {@inheritDoc} */
+    @Override protected void body() throws InterruptedException, IgniteInterruptedCheckedException {
+        while (true) {
+            try {
+                curTask = updateQueue.take();
+            }
+            catch (InterruptedException ignore) {
+            }
 
-                        metastorage.remove(cleanupGuardKey());
-                    }
-                    else if (update instanceof Long) {
-                        long ver = (Long)update;
+            curTask.run();
```

Comment:
As I can see if we catching InterruptedException then curTask may be ``null``, maybe it's better to add one big try/catch?

Reply:
Since ``curTask`` currently is a class field we can get NPE or run the previous task twice (if catching InterruptedException) and it doesn't look good too.

Reply:
Fixed.